### DISCUSSION TITLE
SD Card Mystery

### DISCUSSION THREAD

**Discussion by Kabron287** [earlephilhower/arduino-pico#1904] [General]
Anybody could explain me, how this code from Datalogger example could work:

const int _CS   = 22;
**//#include <SPI.h>**
#include <SD.h>
void setup() {
  // Open serial communications and wait for port to open:
  Serial.begin(115200);
  while (!Serial);
  Serial.print("Initializing SD card...");
  // Ensure the SPI pinout the SD card is connected to is configured properly
  **//SPI1.setRX(_MISO);
  //SPI1.setTX(_MOSI);
  //SPI1.setSCK(_SCK);**
  // see if the card is present and can be initialized:
  if (!SD.begin(_CS)) {
    Serial.println("Card failed, or not present");
    // don't do anything more:
    return;
  }
  Serial.println("card initialized.");

Yes, it successfuly compiled and works onboard with the folliwing pins:
MISO = 16; 
MOSI = 19; 
SCK  = 18;

**Comment by maxgerhardt** [comment]
So you have this code and it works but you don't know why? 

>Yes, it successfuly compiled and works onboard with the folliwing pins:
>MISO = 16;
>MOSI = 19;
>SCK = 18;

Yeah that's great and matches the `SPI` pin map for a Pico.

https://github.com/earlephilhower/arduino-pico/blob/c4f36170d4858b8115e50e2a0caba7c875f1751c/variants/rpipico/pins_arduino.h#L17-L20

Not sure what all the commented out code for SPI1 is about since, if you call `SD.begin(_CS)`, it uses `SPI` (spi0). If you want to use SPI1 instead (with some different pin mappings), you also have to pass `SPI1` as the SPI object in the begin call.

https://github.com/earlephilhower/arduino-pico/blob/c4f36170d4858b8115e50e2a0caba7c875f1751c/libraries/SD/src/SD.h#L34-L41

**Comment by Kabron287** [comment]
BUT I did not defined SPI pins at all.

**Reply by maxgerhardt** [reply]
When you do not call `SPI.setRX()` etc., it will use the default pins as defined in the `pins_arduino.h` file of the selected board variant. And those very clearly say for SPI (0) that MISO = 16, MOSI = 19, SCK = 18. You can see these macros being used to initialize the `SPI` and `SPI1` objects in

https://github.com/earlephilhower/arduino-pico/blob/c4f36170d4858b8115e50e2a0caba7c875f1751c/libraries/SPI/src/SPI.cpp#L366-L372
